feat(azuredevops): support disabled Repos

mr-ks opened the pull request:
### Summary
It is now possible to select disabled repositories as a remote scope. However, disabled repositories are not included in tasks related to domain `code` and `code review`, as the Azure DevOps API returns a 4xx status code for the required API calls.

A new column called `is_disabled` has been added to the `_tool_azuredevops_go_repos` table to support disabled repositories. For existing entries, the default value for this column is set to `false`.
### Does this close any open issues?
Closes #7504

mr-ks committed:
feat(azuredevops): support disabled Azure DevOps Repos

It is now possible to select disabled repositories as a remote scope. However, disabled repositories are not included in tasks related to domain `code` and `code review`, as the Azure DevOps API returns a 4xx status code for the required API calls.

A new column called `is_disabled` has been added to the `_tool_azuredevops_go_repos` table to support disabled repositories. For existing entries, the default value for this column is set to `false`.

mr-ks committed:
fix(azuredevops): harmonize the DomainTypes of the plugin tasks

mr-ks committed:
feat(azuredevops): add support for disabled Azure DevOps repos

mr-ks commented on backend/plugins/azuredevops_go/tasks/pr_commit_collector.go:
```diff
@@ -39,7 +39,7 @@ var CollectApiPullRequestCommitsMeta = plugin.SubTaskMeta{
 	EntryPoint:       CollectApiPullRequestCommits,
 	EnabledByDefault: true,
 	Description:      "Collect PullRequestCommits data from Azure DevOps API.",
-	DomainTypes:      []string{plugin.DOMAIN_TYPE_CROSS, plugin.DOMAIN_TYPE_CODE_REVIEW},
```
Hello @klesh,
what should be the expected outcome if the user creates a scope configuration with only `Cross Domain` as the data entity? Should this also include the tasks of domain `Code Review` and `Source Code Management`?

klesh replied:
"Code review" should be included alongside "Cross Domain".
"Cross Domain" means it would produce "accounts" records as a byproduct.
"Source Code Management" should not be included because it represents the `git repo` code base itself.
"Domain Types" are meant to select which subtasks to run but not limit what they could produce. It is simpler this way and most likey more desirable for users.

mr-ks replied:
Thanks for clarifying. If we classify both the `collect-,convert-,extractAccounts` tasks and the `pullRequests` tasks as `Cross Domain`, we will face issues selecting the correct tasks for disabled ADO repos (and GH repos). Collecting Accounts will work fine, but collecting pull requests will fail. What are your thoughts on this?

Maybe something like this?

```
entities := scopeConfig.Entities without Cross Domain

tasks := helper.MakePipelinePlanSubtasks(subtaskMetas, entities)
tasks = append(tasks, collect-,convert-,extractAccounts)
```

klesh replied:
@mr-ks Good question, I think we could use the [IgnoreAndContinue](https://github.com/apache/incubator-devlake/blob/6dca288d5e33bb55a8b98995df97ead71b76a082/backend/plugins/jenkins/tasks/stage_collector.go#L116) technique for certain HTTP status codes as a workaround.

mr-ks replied:
@klesh  can you please check again? I've integrated your suggestions.

klesh commented on backend/plugins/azuredevops_go/api/blueprint_v200.go:
```diff
@@ -71,52 +69,39 @@ func makeScopeV200(
-		if azuredevopsRepo.Type != models.RepositoryTypeADO {
+		repo, scopeConfig := scope.Scope, scope.ScopeConfig
+
+		if len(scopeConfig.Entities) == 0 {
```
Currently, we are treating this condition as Selected All because it has no point to collect a scope without any entity.

mr-ks replied:
Thanks for the hint, fixed.

klesh commented on backend/plugins/azuredevops_go/api/blueprint_v200.go:
```diff
@@ -126,62 +111,91 @@ func makePipelinePlanV200(
-		var err errors.Error
+		repo, scopeConfig := scope.Scope, scope.ScopeConfig
+
+		if len(scopeConfig.Entities) == 0 {
```
scopeConfig.Entities are considered as "All Entities".

mr-ks replied:
Thanks for the hint, fixed.

klesh commented on backend/plugins/azuredevops_go/api/blueprint_v200.go:
```diff
@@ -126,62 +111,91 @@ func makePipelinePlanV200(
+			continue
+		}
 
 		options := make(map[string]interface{})
```
We have some utility functions to help generate `tasks` and `scopes` , not sure if it could help but you may want to take a look https://github.com/apache/incubator-devlake/blob/5e3a4d7337ded3b27db1931e074a5a46cb1aaca8/backend/plugins/pagerduty/api/blueprint_v200.go

mr-ks replied:
Thanks for highlighting this. Unfortunately, we can't use it here as we need to select tasks based on the repository status. For example, a disabled Azure DevOps repo won't allow us to query its pull requests.

klesh commented on backend/plugins/azuredevops_go/models/migrationscripts/20246008_support_disabled_repos.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (*disabledRepos) Version() uint64 {
+	return 20240806100000
```
the numbers from the file name and the version are different and both incorrect.

mr-ks replied:
Fixed and updated the date to today.

mr-ks committed:
fix(azuredevops): fix inconsistent file name

mr-ks committed:
fix(azuredevops): revert empty entities logic

Empty entities are not treated as 'Selected All' since collecting a scope without any entity is pointless

mr-ks committed:
fix(azuredevops): fix broken test

mr-ks committed:
fix(azuredevops): add support for domain type cross

d4x1 committed:
Merge branch 'main' into feat/ado-disabled-repos

klesh reviewed (approved):
LGTM, thanks for your contribution
